### PR TITLE
T26: MyList<T>

### PR DESCRIPTION
**Pull request by Ashish02112001**

### CODE REVIEW

**Review by ravin-imm — CHANGES_REQUESTED** [review]

File: Program.cs [hunk 1]
```diff
@@ -1,7 +1,135 @@
...
+         }
+      }
+      #endregion
+      #region Methods---------------------------------------------
+      /// <summary>Adds a specified element to the list</summary>
+      /// <param name="a">Element to be added inside the list</param>
+      public void Add (T a) {
+         if (mSize == mElements.Length) {
+            Array.Resize (ref mElements, mElements.Length * 2);
+         }
```
Comment (ravin-imm) [inline Program.cs:92]: Single line after an _if_ statement need not have braces.

File: Program.cs [hunk 2]
```diff
@@ -1,7 +1,135 @@
...
+      }
+      /// <summary>Removes a specified element from the list</summary>
+      /// <param name="a">Element to be removed from the list</param>
+      /// <returns>Returns true if the element is present in the list otherwise false</returns>
+      public bool Remove (T a) {
+         if (mElements.Contains (a)) {
+            for (int i = Array.IndexOf (mElements, a); i < mSize - 1; i++) mElements[i] = mElements[i + 1];
+            mElements[mSize-- - 1] = default;
+            return true;
+         } else return false;
```
Comment (ravin-imm) [inline Program.cs:103]: No need of have _else_ here since this line will be executed if the condition is not valid.

File: Program.cs [hunk 3]
```diff
@@ -1,7 +1,135 @@
...
+      public void RemoveAt (int index) {
+         if (Remove (mElements[index])) return;
+         else throw new IndexOutOfRangeException();
+      }
+      #endregion
+      T[] mElements;
+      int mSize;
    }
+   #endregion
+   #endregion
```
Comment (ravin-imm) [inline Program.cs:134]: _#endregion_ should be placed at appropriate level.

File: Program.cs [hunk 4]
```diff
@@ -1,7 +1,135 @@
...
+         mElements = new T[4];
+         mSize = 0;
+      }
+      /// <summary>Inserts a given element at a specified index</summary>
+      /// <param name="index">Index at which the element has to be insereted</param>
+      /// <param name="a">Element to be inserted in the list</param>
+      public void Insert (int index, T a) {
+         if (mSize == mElements.Length) {
+            Array.Resize (ref mElements, mElements.Length * 2);
+         }
```
Comment (ravin-imm) [inline Program.cs:117]: This line of code that resizes the array is repeated at multiple places. Move this to a method.

File: Program.cs [hunk 5]
```diff
@@ -1,7 +1,135 @@
...
+         if (index > mSize) throw new IndexOutOfRangeException ();
+         mSize++;
+         for (int i = mSize - 1; i >= index + 1; i--) mElements[i] = mElements[i - 1];
+         mElements[index] = a;
+      }
+      /// <summary>Removes the element at the specified index</summary>
+      /// <param name="index">Index at which the element has to be removed</param>
+      public void RemoveAt (int index) {
+         if (Remove (mElements[index])) return;
+         else throw new IndexOutOfRangeException();
```
Comment (ravin-imm) [inline Program.cs:127]: _else_ not required here.

**Review by ravin-imm — CHANGES_REQUESTED** [review]

File: Program.cs [hunk 6]
```diff
@@ -1,7 +1,133 @@
...
+         Array.Clear (mElements, 0, mSize);
+         mElements = new T[4];
+         mSize = 0;
+      }
+      /// <summary>Inserts a given element at a specified index</summary>
+      /// <param name="index">Index at which the element has to be insereted</param>
+      /// <param name="a">Element to be inserted in the list</param>
+      public void Insert (int index, T a) {
+         ExpandArraySize ();
+         if (index > mSize) throw new IndexOutOfRangeException ();
```
Comment (ravin-imm) [inline Program.cs:115]: First the exception has to be thrown.

**Review by ravin-imm — CHANGES_REQUESTED** [review]

**Comment by ravin-imm** [comment]
I don't see any tests in your last commit. Please add tests.

File: Training/Program.cs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+         }
+         set => mElements[index] = value;
+      }
+      #endregion
+      #region Methods -----------------------------------------------
+      /// <summary>Adds a specified element to the list</summary>
+      /// <param name="a">Element to be added inside the list</param>
+      public void Add (T a) {
+         mElements[mSize++] = a;
+         ExpandArraySize ();
```
Comment (ravin-imm) [inline Training/Program.cs:92]: I think the better design is to first check if the array can accommodate the number to be added. If not, expand the array and then add the element. This way we will be expanding the array only if we are adding once it is full. In the above implementation, we add an element and try to expand the array if it gets full.

File: TestProject1/UnitTest1.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+         Assert.ThrowsException<IndexOutOfRangeException> (() => myNums.RemoveAt (8));
+         nums.Clear (); myNums.Clear ();
+      }
+      [TestMethod]
+      public void TestCapacity () {
+         nums.Add (1); myNums.Add (1);
+         nums.Add (2); myNums.Add (2);
+         nums.Add (3); myNums.Add (3);
+         nums.Add (4); myNums.Add (4);
+         Assert.AreEqual (8, myNums.Capacity);
```
Comment (ravin-imm) [inline TestProject1/UnitTest1.cs:62]: This is not a valid test. Though the array has only 4 elements, the capacity has been doubled.

**Review by ravin-imm — CHANGES_REQUESTED** [review]

File: Training/Program.cs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            else return mElements[index];
+         }
+         set => mElements[index] = value;
+      }
+      #endregion
+      #region Methods -----------------------------------------------
+      /// <summary>Adds a specified element to the list</summary>
+      /// <param name="a">Element to be added inside the list</param>
+      public void Add (T a) {
+         if (mSize >= Capacity) ExpandArraySize ();
```
Comment (ravin-imm) [inline Training/Program.cs:91]: There is a double check here for the array size. This check is done in _ExpandArraySize_ () method also.

File: TestProject1/UnitTest1.cs [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+      List<int> nums = new ();
+      MyList<int> myNums = new ();
+
+      [TestMethod]
+      public void TestAdd () {
+         nums.Add (1); myNums.Add (1);
+         nums.Add (2); myNums.Add (2);
+         nums.Add (3); myNums.Add (3);
+         Assert.AreEqual (nums[0], myNums[0]);
+         nums[0] = 0; myNums[0] = 0;
```
Comment (ravin-imm) [inline TestProject1/UnitTest1.cs:14]: Why is the special treatment for the first element here?

**Review by ravin-imm — CHANGES_REQUESTED** [review]

File: TestProject1/UnitTest1.cs (hunk 10 above)
Comment (Ashish02112001) [reply TestProject1/UnitTest1.cs:14]: Just for testing the set by changing the value sir.

**Pull request closed without merging**